client-spark/common/src/main/scala/org/apache/spark/serializer/ForySerializer.scala

```diff
@@ -0,0 +1,181 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.serializer
+
+import org.apache.fory.Fory
+import org.apache.fory.config.{CompatibleMode, Language}
+import org.apache.spark.internal.Logging
+
+import java.io.{InputStream, OutputStream, Serializable}
+import java.nio.ByteBuffer
+import scala.reflect.ClassTag
+
+class ForySerializer extends org.apache.spark.serializer.Serializer
+  with Logging
+  with Serializable {
+
+  override def newInstance(): SerializerInstance = new ForySerializerInstance()
+
+  override def supportsRelocationOfSerializedObjects: Boolean = true
+
+}
+
+class ForySerializerInstance extends org.apache.spark.serializer.SerializerInstance {
+
+  // Thread-local Fury instance for thread safety
+  private val fury: ThreadLocal[Fory] = ThreadLocal.withInitial(() => {
+    val f = Fory.builder()
+      .withLanguage(Language.JAVA)
+      .withRefTracking(true)
+      .withCompatibleMode(CompatibleMode.COMPATIBLE)
+      .requireClassRegistration(false)
+      .build()
+    f
+  })
+
+  override def serialize[T: ClassTag](t: T): ByteBuffer = {
+    val bytes = fury.get().serialize(t.asInstanceOf[AnyRef])
+    ByteBuffer.wrap(bytes)
+  }
+
+  override def deserialize[T: ClassTag](bytes: ByteBuffer): T = {
+    val array = if (bytes.hasArray) {
+      val offset = bytes.arrayOffset() + bytes.position()
+      val length = bytes.remaining()
+      java.util.Arrays.copyOfRange(bytes.array(), offset, offset + length)
+    } else {
+      val array = new Array[Byte](bytes.remaining())
+      bytes.get(array)
+      array
+    }
+    fury.get().deserialize(array).asInstanceOf[T]
+  }
+
+  override def deserialize[T: ClassTag](bytes: ByteBuffer, loader: ClassLoader): T = {
+    // Fury handles class loading internally, so we can use the standard deserialize method
+    deserialize[T](bytes)
+  }
+
+  override def serializeStream(s: OutputStream): SerializationStream = {
+    new ForySerializationStream(fury.get(), s)
+  }
+
+  override def deserializeStream(s: InputStream): DeserializationStream = {
+    new ForyDeserializationStream(fury.get(), s)
+  }
+}
+
+class ForySerializationStream(fury: Fory, outputStream: OutputStream)
+  extends org.apache.spark.serializer.SerializationStream {
+
+  private val out = outputStream
+  private var closed = false
+
+  override def writeObject[T: ClassTag](t: T): SerializationStream = {
+    if (closed) {
+      throw new IllegalStateException("Stream is closed")
+    }
+    
+    val bytes = fury.serialize(t.asInstanceOf[AnyRef])
```

Comment:
Maybe hold a Fory `MemoryBuffer` as an instance field in the class, and serialize object into that buffer, then you can get heap buffer from that buffer, and write it into `out`. In this way, yo u can reduce a copy